content/4_Gold/LIS.mdx

```diff
@@ -338,6 +342,68 @@ class PCB{
 
 </LanguageSection>
 
+## Extension - LCS on Permutations
+
+Given two permutations of length $N$ ($1\le N \le 10^5$), find the longest common subsequence.
+
+### Slow Solution
+
+**Time Complexity**: $\mathcal O(N^2)$
+
+See [here](/gold/paths-grids#solution---longest-common-subsequence).
+
+### Efficient Solution
+
+**Time Complexity**: $\mathcal O(N\log N)$
+
+In the $\mathcal O(N^2)$ solution, we increase the DP arrray only if $a[i] = b[j]$, but since both arrays are permutations of length $N$, for each $a[i]$, there must be only one matching element in $b$.  
+
+Let us create an array $c$ where $c[i]$ is $a^{-1}[b[i]]$. Our final LIS $c[x_1], c[x_2], \dots c[x_k]$ corresponds to the LCS $b[x_1], b[x_2], \dots b[x_k]$, which is equivalent to $a[c[x_1]], a[c[x_2]], \dots a[c[x_k]]$.
+
+In other words, every increasing subsequence $x_{1 \dots k}$ in $c$ corresponds to a common subsequence between $a$ and $b$.
+
+Finding a LIS takes $\mathcal O(N\log N)$, so the overall time complexity in this algorithm is $\mathcal O(N\log N)$.
+
+<LanguageSection>
+
+<CPPSection>
+
+```cpp
+#include <bits/stdc++.h>
+using namespace std;
+
+const int N = 1e5 + 5;
+int n, a[N], pos[N], b[N];
+vector<int> lis;
+
+int main() {
+	cin.tie(0)->sync_with_stdio(0);
+
+	cin >> n;
+	for (int i = 1; i <= n; ++i) {
+		cin >> a[i];
+		pos[a[i]] = i;
+	}
+	for (int i = 1; i <= n; ++i) {
+		cin >> b[i];
+		b[i] = pos[b[i]];
+	}
+	for (int i = 1; i <= n; ++i) {
+		int p = lower_bound(begin(lis), end(lis), b[i]) - begin(lis);
+		if (p == lis.size())
+			lis.push_back(b[i]);
+		else
+			lis[p] = b[i];
+	}
+	cout << lis.size() << '\n';
+}
+```
+
+</CPPSection>
```

Comment:
Would be nice if you explicitly created array c in the code as well. If not, add a comment saying smth along the lines of "instead of creating permutation $c$, we modify permutation b"